pypesto/objective/amici_prediction.py

```diff
@@ -0,0 +1,435 @@
+            Default is 0 meaning that all conditions will be simulated.
+            Other values are only applicable, if an output file is specified.
+        observables:
+            IDs of observables, if post-processing is used
+        """
+        # save settings and objective
+        self.amici_objective = amici_objective
+        self.max_num_conditions = max_num_conditions
+        self.post_processing = post_processing
+        self.post_processing_sensi = post_processing_sensi
+        self.post_processing_timepoints = post_processing_timepoints
+
+        if observables is None:
+            self.observables = amici_objective.amici_model.getObservableIds()
+        else:
+            self.observables = observables
+
+    def __call__(
+            self,
+            x: np.ndarray,
+            sensi_orders: Tuple[int, ...] = (0, ),
+            mode: str = MODE_FUN,
+            output_file: str = '',
+            output_format: str = 'csv',
+            max_num_conditions: int = 0,
+    ) -> PredictionResult:
+        """
+        Simulate a model for a certain prediction function.
+        This method relies on the AmiciObjective, which is underlying, but
+        allows the user to apply any post-processing of the results and the
+        sensitivities.
+
+        Parameters
+        ----------
+        x:
+            The parameters for which to evaluate the prediction function.
+        sensi_orders:
+            Specifies which sensitivities to compute, e.g. (0,1) -> fval, grad.
+        mode:
+            Whether to compute function values or residuals.
+        output_file:
+            Path to an output file.
+        max_num_conditions:
+            In some cases, we don't want to compute all predictions at once
+            when calling the prediction function, as this might not fit into
+            the memory for large datasets and models.
+            Here, the user can specify a maximum number of conditions, which
+            should be simulated at a time. Default is self.max_num_conditions.
+        output_format:
+            Either 'csv', 'h5'. If an output file is specified, this routine
+            will return a csv file, created from a DataFrame, or an h5 file,
+            created from a dict.
+
+        Returns
+        -------
+        outputs:
+            List of postprocessed outputs (ndarrays)
+        outputs_sensi:
+            List of sensitivities of postprocessed outputs (ndarrays)
+        """
+        # sanity check for output
+        if 2 in sensi_orders:
+            raise Exception('Prediction simulation does currently not support '
+                            'second order output.')
+
+        # prepare for storing the results
+        amici_y = []
+        amici_sy = []
+        amici_t = []
+
+        # simulate the model and get the output
+        self._get_model_outputs(amici_y, amici_sy, amici_t, x,
+                                sensi_orders, mode)
```

Comment:
Is preparing the results objects here useful? `_get_model_outputs` could return a `dict` (created within `_get_model_outputs`, not here) with keys `{'t', 'y', 'sy'}`. This might make it easier to extend for states later too

Reply:
If done then post-processors could also be specified with a `dict` (would simplify some code a bit).

Reply:
One could do that via a dict. I just see that this function has become something different than I originally had in mind... hm... In the way it looks now, preparation makes no sense any more. Will change that, because the encapsulated call to the actual simulation is important for me.